IdentityCore/src/webview/MSIDWebviewAuthorization.m

```diff
@@ -26,33 +26,213 @@
+                                                     context:(id<MSIDRequestContext>)context
+                                           completionHandler:(MSIDWebUICompletionHandler)completionHandler
 {
-    return nil;
+    id<MSIDWebviewInteracting> embeddedWebviewController = [factory embeddedWebviewControllerWithRequest:parameters
+                                                                                           customWebview:webview
+                                                                                                 context:context
+                                                                                       completionHandler:completionHandler];
+    [self startWebviewAuth:embeddedWebviewController
+                   context:context
+         completionHandler:completionHandler];
 }
 
-+ (id<MSIDWebviewInteracting>)embeddedWebviewControllerWithRequestParameters:(MSIDRequestParameters *)parameters
-                                                                     webview:(WKWebView *)webview
-                                                                     factory:(MSIDOAuth2Factory *)factory;
++ (void)startSystemWebviewWebviewAuthWithRequestParameters:(MSIDRequestParameters *)parameters
+                                                   factory:(MSIDOauth2Factory *)factory
+                                                   context:(id<MSIDRequestContext>)context
+                                         completionHandler:(MSIDWebUICompletionHandler)completionHandler
 {
+
+    id<MSIDWebviewInteracting> systemWebviewController = [factory systemWebviewControllerWithRequest:parameters
+                                                                                   callbackURLScheme:parameters.redirectUri
+                                                                                             context:context
+                                                                                   completionHandler:completionHandler];
+    [self startWebviewAuth:systemWebviewController
+                   context:context
+         completionHandler:completionHandler];
+}
+
++ (void)startWebviewAuth:(id<MSIDWebviewInteracting>)webviewController
+                 context:(id<MSIDRequestContext>)context
+       completionHandler:(MSIDWebUICompletionHandler)completionHandler
+{
+    if (![self setCurrentWebSession:webviewController])
+    {
+        NSError *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInteractiveSessionAlreadyRunning, @"Only one interactive session is allowed at a time.", nil, nil, nil, context.correlationId, nil);
+        
+        completionHandler(nil, error);
+        return;
+    }
+    
+    if (![s_currentWebSession start])
+    {
+        NSError *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInteractiveSessionStartFailure, @"Interactive web session failed to start.", nil, nil, nil, context.correlationId, nil);
+        
+        completionHandler(nil, error);
+    }
```

Comment:
Since we have already passed `completionHandler` to system/embedded webview, shall we let the webview class to call the completion block if it fails to start?
The `start` action is kind of async because it involves UI/main thread, it may not be able to return a value immediately to indicate if the start action succeeds :)

Reply:
Discussed with JK offline, embedded webview will return YES for the start method, to avoid the completion block being called twice :)